### DISCUSSION TITLE
How do I track changes inside a tracked directory?

### DISCUSSION THREAD

**Discussion by plucafs** [twpayne/chezmoi#3988] [Q&A] [answered]
Hello!
With chezmoi I'm trying to keep track of the files inside `~/.config/fish`. When I first added the directory, pushed and then runned `chezmoi  update` on my second machine, all went fine. 

The problem is when I add/remove a file inside `~/.config/fish/functions`, as an example.
When I execute `chezmoi re-add` or `chezmoi add "~/.config/fish"` nothing changes.

I was expecting chezmoi to keep track of the added/removed files, since the directory that I'm tracking is `~/.config/fish`. Probably I'm not understanding correctly how this process works.
Thanks!

**Comment by twpayne** [comment]
By default, chezmoi will only manage entries that are explicitly listed in the source state, i.e. it will not remove unmanaged files.

For cases like yours when you want chezmoi to manage the exact contents of a directory, including removing unmanaged files, you need to use the [`exact_` attribute])https://www.chezmoi.io/reference/target-types/#directories).

```console
$ chezmoi chattr +exact ~/.config/fish/functions
```

If you use `exact_` directories, then be sure to run `chezmoi diff` before `chezmoi apply` so you can check exactly what files `chezmoi apply` will remove.

**Reply by plucafs** [reply]
Ok, adding the `+exact` attribute synced the content of the directory on the second machine.

But, I'm not sure if I'm using the correct workflow, since when I add a file in `.config/fish/functions`, I need to use `chezmoi add .config/fish/functions` to make chezmoi push the changes. And when I delete the same file, neither `chezmoi add .config/fish/functions` and `chezmoi re-add` works. 

To give more context, I never use `chezmoi cd` and I always edit the files in the source directory, and use `chezmoi re-add`, with auto commit and auto push enabled.

Thanks!

**Reply by twpayne** [reply, marked as answer]
> And when I delete the same file, neither `chezmoi add .config/fish/functions` and `chezmoi re-add` works.

That is correct. You need to use `chezmoi forget` or `chezmoi destroy`.



> To give more context, I never use `chezmoi cd` and I always edit the files in the source directory, and use `chezmoi re-add`, with auto commit and auto push enabled.

You should use `chezmoi cd`. `chezmoi re-add` does not handle the case you want (although this might change in the future, see #2298).

**Reply by plucafs** [reply]
Works correctly. Thanks for the fast replies and the support!
